lib/evaluate/rte-interface.h

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2018-2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_EVALUATE_RTE_INTERFACE_H_
+#define FORTRAN_EVALUATE_RTE_INTERFACE_H_
+
+// Defines the structure that must be used in F18 when dealing with the
+// runtime, either with the target runtime or with the host runtime for
+// folding purposes
+// To avoid unnecessary header circular dependencies, the actual implementation
+// of the templatized member function are defined in rte.h
+// The header at hand must be included in order to add the rte interface data
+// structure as a member of some structure.
+// To actually use the rte interface, rte.h must be included. Note that rte.h
+// includes the header at hand.
+
+#include <functional>
+#include <map>
+#include <optional>
+#include <string>
+#include <vector>
+
+namespace Fortran::evaluate {
+class FoldingContext;  // forward declare
+namespace rte {
+
+using TypeCode = unsigned char;
```

Comment:
Are you using single character encodings of types rather than an enumeration or (category,kind) ordered pair?

Are there mapping functions to encode/decode these single character type codes?

Reply:
The actual mapping is defined in rte.h. see typeCodeOf<T> and TypeOf<TypeCode n>.
This is both to avoid dependencies and also because I do not really want to make it possible for the user to interact with the type encoding directly.
TypeCode is simply the index of a type inside the tuple `RteTypes` containing all the types that are allowed to describe a runtime function symbol. For now RteTypes are all the intrinsic types.
I do not want to define an enum for now because it will be annoying define and maintain.
I did not consider using pairs (category, kind) because I had the void and descriptor types in mind that do not really fit inside this encoding and are special. It is easy to just add an empty class DescriptorType and VoidType to reflect that.
I tried to centralized everything interacting with this encoding so that it can "mature" when trying to describe more complex runtime functions.